hawc/apps/common/management/commands/clean_doi.py

```diff
@@ -0,0 +1,158 @@
+        doi_id_initial = doi_identifiers.count()
+        validate_dois(doi_identifiers)
+        refs = Reference.objects.exclude(identifiers__database=4)
+        doi_ref_initial = Reference.objects.filter(identifiers__database=4).count()
+        if options["extensive"]:
+            create_dois_extensive(refs)
+        else:
+            create_dois(refs)
+        doi_id_final = Identifiers.objects.filter(database=4).count()
+        doi_ref_final = Reference.objects.filter(identifiers__database=4).count()
+        ref_total = Reference.objects.count()
+        print(f"Began with {doi_id_initial} doi IDs; finished with {doi_id_final} doi IDs")
+        print(
+            f"Began with {doi_ref_initial} references with doi IDs; finished with {doi_ref_final} references with doi IDs, out of {ref_total} total references"
+        )
+
+
+def validate_dois(doi_identifiers):
+    print(f"Validating {doi_identifiers.count()} doi IDs...")
+    doi_id_saved = {}
+    doi_id_updated = []
+    for doi_id in doi_identifiers:
+        if constants.DOI_EXACT.fullmatch(doi_id.unique_id):
+            doi_id_saved[doi_id.unique_id] = doi_id
+            # doi_id_saved.append(doi_id.unique_id)
+
+    duplicate_doi_ids = {}  # stores unclean DOIs that have a clean duplicate for updating later
+    tarnished_doi_ids = []  # tarnished DOI IDs cannot be cleaned/validated (ex: '10')
+    for doi_id in doi_identifiers:
+        # converts html character references to actual Unicode characters (ex: &gt; to >)
+        new_doi = html.unescape(doi_id.unique_id)
+        if constants.DOI_EXACT.fullmatch(new_doi):
+            if new_doi not in doi_id_saved.keys():
+                # html conversion made a difference; update
+                doi_id.unique_id = new_doi
+                doi_id_updated.append(doi_id)
+                doi_id_saved[new_doi] = doi_id
+            elif new_doi != doi_id.unique_id:
+                # html conversion made a difference but theres an existing duplicate
+                other_dupes = duplicate_doi_ids.get(new_doi, [])
+                other_dupes.append(doi_id)
+                duplicate_doi_ids[new_doi] = other_dupes
+
+        else:
+            # doi needs to be further cleaned/validated
+            new_doi = constants.DOI_EXTRACT.search(new_doi)
+            if new_doi:
+                new_doi = new_doi.group(0)
+            if new_doi and new_doi.endswith("."):  # remove period at end of doi if it exists
+                new_doi = new_doi[:-1]
+            if new_doi is None or not constants.DOI_EXACT.fullmatch(new_doi):
+                # extraction failed; doi is tarnished
+                tarnished_doi_ids.append(doi_id)
+            else:
+                if new_doi in doi_id_saved.keys():
+                    other_dupes = duplicate_doi_ids.get(new_doi, [])
+                    other_dupes.append(doi_id)
+                    duplicate_doi_ids[new_doi] = other_dupes
```

Comment:
use [defaultdict](https://docs.python.org/3/library/collections.html#collections.defaultdict); `duplicate_doi_ids=defaultdict(list)` and then you just `duplicate_doi_ids[key].append(value)`